src/libraries/System.Net.Http/src/System/Net/Http/HttpMessageInvoker.cs

```diff
@@ -100,11 +104,11 @@ static async Task<HttpResponseMessage> SendAsyncWithTelemetry(HttpMessageHandler
             request.RequestUri is Uri requestUri &&
             requestUri.IsAbsoluteUri;
 
-        internal static bool LogRequestFailed(bool telemetryStarted)
+        internal static bool LogRequestFailed(Exception exception, bool telemetryStarted)
         {
             if (HttpTelemetry.Log.IsEnabled() && telemetryStarted)
             {
-                HttpTelemetry.Log.RequestFailed();
+                HttpTelemetry.Log.RequestFailed(exception.ToString());
```

Comment:
I thought the request was for the exception type name rather than the entire exception string, did I misunderstand? Exception.ToString() could be quite large and would likely need to be parsed to be useful which makes it a bit unwieldy.

Reply:
Also, is cardinality important to events like this? Eg if the exception message contains a component that varies, whatever processes these events would not be able to easily aggregate and summarize them.

Reply:
With HttpClient the exception type on its own is not useful (`HttpRequestException` or `TaskCanceledException`).
The message is the bare minimum to get actual information in some cases (e.g., was it canceled vs did it time out vs response headers were too large).

But even then it might be useless on its own without the inner exception and/or stack trace.
For example:
```
System.Threading.Tasks.TaskCanceledException: The operation was canceled.
 ---> System.IO.IOException: Unable to read data from the transport connection: The I/O operation has been aborted because of either a thread exit or an application request..
 ---> System.Net.Sockets.SocketException (995): The I/O operation has been aborted because of either a thread exit or an application request.
   --- End of inner exception stack trace ---
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource<System.Int32>.GetResult(Int16 token)
   at System.Net.Security.SslStream.EnsureFullTlsFrameAsync[TIOAdapter](CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.PoolingAsyncValueTaskMethodBuilder`1.StateMachineBox`1.System.Threading.Tasks.Sources.IValueTaskSource<TResult>.GetResult(Int16 token)
   at System.Net.Security.SslStream.ReadAsyncInternal[TIOAdapter](Memory`1 buffer, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.PoolingAsyncValueTaskMethodBuilder`1.StateMachineBox`1.System.Threading.Tasks.Sources.IValueTaskSource<TResult>.GetResult(Int16 token)
   at System.Net.Http.HttpConnection.InitialFillAsync(Boolean async)
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
```
and
```
System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at MyStream.ReadAsync(Memory`1 buffer, CancellationToken cancellationToken)
   at System.IO.Stream.<CopyToAsync>g__Core|27_0(Stream source, Stream destination, Int32 bufferSize, CancellationToken cancellationToken)
   at System.Net.Http.StreamToStreamCopy.<CopyAsync>g__DisposeSourceAsync|1_0(Task copyTask, Stream source)
   at System.Net.Http.HttpContent.<CopyToAsync>g__WaitAsync|56_0(ValueTask copyTask)
   at System.Net.Http.HttpConnection.SendRequestContentAsync(HttpRequestMessage request, HttpContentWriteStream stream, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
...
```
are both the same exception type with the same message, but with a completely different root cause.

Reply:
I think we should probably chat with Jeff to understand what data he is hoping for. If he wants the full exception text I won't object, but I'm skeptical. We might also want to put events that generate that much data under their own keyword because building and formatting the complete Exception text including stack traces and inner exceptions is probably 10-1000x slower than emitting a typical ETW event depending on external factors. One pattern we need to be wary of is where telemetry overhead is low when the service is operating normally, but then some problem like a network issue occurs that causes the service to temporarily but consistently go down an error path for a little while. If this causes the overhead from telemetry to dramatically increase sometimes this will change what would have been a temporary issue into a new [stable self-reinforcing failure mode](http://charap.co/metastable-failures-in-distributed-systems/) that persists independently of the condition which triggered it. The unexpected new load may also have spill-over effects onto other elements of the service and the telemetry system. I know this might sound like a theoretical concern but there was a real production outage of a major Azure service not too long ago that did unfold like that and the culprit which caused the overhead was excessive calls to StackTrace.ToString().

Reply:
Adding a bit to what Noah says here - There is already an event that is fired on thrown exceptions that comes with a call stack generated by the ETW subsystem.  This is something that telemetry consumers can choose to enable if they want it, but can also choose not to enable if they don't.  I think it's valuable to be able to correlate the exception with the request, but that doesn't necessarily have to happen in code in `System.Net`.  I'm wondering if you can do a test where you log that the request failed using the old event format without the exception payload, and see if the activityID is attached to the event.  If it is, then there is already built-in correlation, and you can avoid the call to `Exception.ToString()`.

Reply:
> building and formatting the complete Exception text including stack traces and inner exceptions is probably 10-1000x slower than emitting a typical ETW event ...

@noahfalk is your main concern the performance overhead of materializing the stack trace string?

Doing a rough benchmark, materializing the stack is similarly expensive as throwing the exception in the first place, making the exception path "2x slower". The overhead of an ETW event is noise compared to this.
But that is without any networking logic involved. Seeing an exception with HttpClient often means that you've also just lost a connection, so you're also paying for multiple network round trips, a full TLS handshake, etc. ... Are the 50 extra μs really that impactful in that case?

Do you think it should be behind an opt-in keyword/level, a separate event? Or are you against including the info in an event in general?

> ... culprit which caused the overhead was excessive calls to StackTrace.ToString()

From doing `StackTrace.ToString` on the exception path?

> could be quite large and would likely need to be parsed to be useful which makes it a bit unwieldy

I would be okay with splitting it into two parameters (e.g. `exceptionMessage, exceptionStackTrace`) if you think that'd be easier to consume.

> There is already an event that is fired on thrown exceptions that comes with a call stack generated by the ETW subsystem.

Does that also dump the inner exceptions and stack traces?

We flow the `ExecutionContext` through our logic and have tests verifying that ActivityIds are flowing through. But the networking logic is not always a simple stack down to a call to the network. Consider logic like
```c#
private TaskCompletionSource _tcs;

// Some random background task doing I/O
_tcs.SetException(new Exception("Useful info"));

// Caller logic that flows ActivityIDs that will eventually unwind up to HttpClient.SendAsync
try
{
    await _tcs.Task;
}
catch (Exception ex)
{
   throw new Exception("Something threw", ex);
}
```
would the "Useful info" bit be captured?

Even so, if the concern is performance, is turning on more ETW providers, capturing exceptions for the whole process, and then correlating them really better?

---

I still believe that having the stack trace is useful. If someone comes to us (networking team) for help with the information that they're seeing a lot of `The operation was canceled.` failures, we have to send them back to turn on additional event providers, which come with a much higher performance overhead, and try to reproduce the issue again while capturing more information.
If they come to us with "99% of stack traces share _insert stack frame here_", we're much more likely to be able to provide any useful feedback immediately.

Reply:
> Do you think it should be behind an opt-in keyword/level, a separate event? Or are you against including the info in an event in general?

No qualm having it present as an option, I'd just prefer it is opt-in, presumably using a keyword.

I also suspect that Jeff may be looking for telemetry that has more predictable size and is easier to parse so you might wind up with say Exception type name or message parameters for him use-case and then an optional more complete exception information behind a keyword. I want to make sure we are satisfying Jeff's needs in addition to whatever we are doing to help ourselves.

If you really think it is important to have complete exception info by default then I would strongly suggest avoiding Exception.ToString() so that you can control how the stack trace is formatted and use StackTrace(Exception, needFileInfo:false) to disable all the symbol handling code. That is where a majority of performance risk enters the picture. You'd probably also want limits on nested exception recursion depth and total formatted text length to ensure pathological cases don't generate huge event payloads.


> Doing a rough benchmark, materializing the stack is similarly expensive as throwing the exception in the first place

There are many variables involved in the performance of formatting the stack trace (number of frames, number of modules, presence of symbols, format of symbols, are the symbols embedded, size of module and symbol files, speed of disk IO, number of concurrent threads contending for the locks in this code path, etc). You can measure a not too scary number from a benchmark and then get dramatically worse performance under less favorable conditions. My goal is that if people turn on some default telemetry the overhead is as predictable as possible, including when they happen to be operating in the less favorable conditions.

Reply:
@MihaZupan I want to make sure you saw my note above.  There is already an option to capture ETW events for all thrown exceptions, and this will come with the stack trace, but in an ETW-consumable format instead of just text.  This can be correlated with the `HttpClient` events, and so you can get everything, but the consumer can decide how much data they want.  If they just want to know if a failure occurred, they turn on System.Net.Http events.  If they want the exception name, message, and stacktrace, they turn on the CLR's exception events.  This would also come with the side benefit of not needing to create more configuration (e.g. another keyword) that someone needs to learn about to consume.  What do you think?

Reply:
@brianrob That is what I was referring to in the

> Does that also dump the inner exceptions and stack traces?

section of my reply above.

> you can control how the stack trace is formatted and use StackTrace(Exception, needFileInfo:false) to disable all the symbol handling code. That is where a majority of performance risk enters the picture.

Thanks, I wasn't aware of this option.

Reply:
Thanks @MihaZupan, I missed that.  It does not dump the inner exception and stack trace, but if the inner exception is thrown, then it will have its own event and stack trace.

Reply:
@brianrob - I assume one other meaningful distinction between the stacktrace ETW/EventPipe currently provides for Exception events and Exception.StackTrace is that the latter handles async code the way most users would expect. For async code the stacktraces from ETW/EventPipe can be near useless if they are just showing the threadpool and Task implementation plus a tiny snippet of the async code which ran synchronously.

Reply:
Copying the example from that comment:

> We flow the `ExecutionContext` through our logic and have tests verifying that ActivityIds are flowing through. But the networking logic is not always a simple stack down to a call to the network. Consider logic like
> ```c#
> private TaskCompletionSource _tcs;
> 
> // Some random background task doing I/O
> _tcs.SetException(new Exception("Useful info"));
> 
> // Caller logic that flows ActivityIDs that will eventually unwind up to HttpClient.SendAsync
> try
> {
>     await _tcs.Task;
> }
> catch (Exception ex)
> {
>    throw new Exception("Something threw", ex);
> }
> ```
> would the "Useful info" bit be captured?

The "Useful info" would be logged at some point, but there'd be no way to correlate that back to the exception thrown by HttpClient, right?

Reply:
@noahfalk we discussed the options here a bit more with the team and the majority opinion for now is that we want to expose the inner exception/stack information in the System.Net.Http event source, but we're okay with pushing it to a separate event.

That means we're thinking of adding a new event to include the whole exception info, but under a verbose `EventLevel`.

```c#
[EventLevel.Error]
void RequestFailed(string exceptionMessage);

[EventLevel.Verbose]
void RequestFailedVerbose(string exception);
```
This would be the first verbose event on this provider.

Any concerns with that?

Reply:
> @brianrob - I assume one other meaningful distinction between the stacktrace ETW/EventPipe currently provides for Exception events and Exception.StackTrace is that the latter handles async code the way most users would expect. For async code the stacktraces from ETW/EventPipe can be near useless if they are just showing the threadpool and Task implementation plus a tiny snippet of the async code which ran synchronously.

It's true - the stack is the true native stack, and not one that has been "demangled".

If you really want a separate event with the full exception and stack, making it a verbose event is a good plan.  It looks like all of the other events are explicitly marked with a level, so this would be a safe way to ensure that this doesn't get inadvertently enabled.

Reply:
> This would be the first verbose event on this provider.
Any concerns with that?

I don't think there is any strong pattern around this, but I would suggest using a keyword rather than the verbose level as the mechanism to opt-in. The challenge I've seen in the past with noisy data is that once you have a few different kinds consumers will want to opt-in just to the specific ones they care about. For low-volume data this generally isn't an issue because receiving a few unwanted low-volume events has little overhead, but for potentially high overhead data its worthwhile for consumers to be choosy. Putting all the noisy data into a single 'Verbose' bucket doesn't allow for the fine-grained control. It is possible to put the first piece of noisy data under the Verbose setting and then all future noisy data can be configured as Informational + separate keyword bits, but the asymmetry may make the design feel slightly inconsistent to consumers if you ever reach the point of both verbose and keyword bits in use.

Reply:
So keep the new detailed event on the same `EventLevel.Error` under a keyword opt-in? Sounds good to me

Reply:
I would recommend setting the Level to `Verbose`, but also put it behind a keyword so that you can control which verbose events you want by selecting the right keyword.  Fundamentally, this is a costly event, so you'd want it to be treated as verbose.

Reply:
> are both the same exception type with the same message, but with a completely different root cause.

in this case it looks like you could differentiate by just concatenating the messages of inner exceptions. (that has to be done manually, I'm not aware of a method to do this). .. is that generally sufficient?

Reply:
It'd be better than just the top-level message, but I don't think adding extra logic makes sense in this case over just serializing the stack via the built-in helper.

> I would recommend setting the Level to Verbose, but also put it behind a keyword so that you can control which verbose events you want by selecting the right keyword. Fundamentally, this is a costly event, so you'd want it to be treated as verbose.

Doesn't that defeat Noah's comment though? It would mean that any new verbose events we add would need separate keywords, or we'd make collecting this one even more expensive.
I could see myself just turning on `EventLevel.Error/Info` with a keyword to request additional error information, without caring about verbose events.

Reply:
@brianrob - Mostly agreeing with Miha here... assuming that we already have the keyword, is requiring verbose level giving a benefit? In the best case scenario I am imagining we are diligent and every verbose event we add has a corresponding keyword so that users can opt into exactly the subset of expensive events they want. In this model the verbose marking doesn't seem to help or hurt. In the worst case we have a bunch of verbose events with no keyword that allows them to be filtered out and now anyone who wants this data also needs to receive all the other verbose events even if they have no use for that portion of it. Here is an [example](https://github.com/dotnet/runtime/issues/11775) of that problem playing out in the past. Mukul wanted some JIT info that required both a keyword and the verbose level, but setting level=verbose opted him into receiving other verbose GC info he didn't want.